### PR TITLE
Auditlog cleanup and update

### PR DESCRIPTION
**Pull request by Plerx2493**
# Summary
I started to cleanup the GetAuditLogs method on DiscordGuild and updated the AuditLogActionType Enum

# Notes
Feel free to give me feedback and point out additional methods i can cleanup in this scope. I thought about rewriting other auditlog related sections but i dont know entirely how it works

# Tasks
- [x] Implement missing entry types
- [x] Implement parsing of missing types
- [x] Implement missing AuditLog event

### CODE REVIEW

**Comment by Plerx2493** [comment]
i have not implemented a new type for the new actionTypes because i didnt know if we want to continue with this monster method

**Comment by Plerx2493** [comment]
The editorconfig is against using var. Should i replace it in all methods i clean up? its used everywhere

**Comment by akiraveliara** [comment]
please do, and if you can, also run other format rules like file-scoped namespaces

**Comment by Plerx2493** [comment]
@akiraveliara should i enforce all editorconfig warnings or only those which only affects files i already have changes in?

**Comment by Plerx2493** [comment]
There are several naming rules broken but they also make changes in DiscordClient.cs

**Comment by akiraveliara** [comment]
only in files you touch for now, so as to avoid huge merge conflicts

**Comment by Plerx2493** [comment]
ok i got it

**Comment by Plerx2493** [comment]
Now its a huge PR for a minor change ^^

**Comment by akiraveliara** [comment]
you can always make the PR extend to more audit log functionality if you want (:

**Comment by akiraveliara** [comment]
but seriously, that's fine. it'll be a little annoying when it comes to reverting commits, but eh

**Comment by Plerx2493** [comment]
My plan is to implement the missing auditlogtypes. I will see what I can do after that

**Comment by Plerx2493** [comment]
- [ ] Implement missing entry types
- [ ] Implement instanciation when fetching
- [ ] Implement missing AuditLog event

**Comment by Plerx2493** [comment]
Parsing is not fully tested (We agreed on making a testing post), new event is tested.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 1]
```diff
@@ -1214,6 +1221,16 @@ internal async Task OnGuildIntegrationsUpdateEventAsync(DiscordGuild guild)
             };
             await this._guildIntegrationsUpdated.InvokeAsync(this, ea);
         }
+        
+        private async Task OnGuildAuditLogEntryCreateEventAsync(DiscordGuild guild, DiscordAuditLogEntry toDiscordObject)
```
Comment (akiraveliara) [inline DSharpPlus/Clients/DiscordClient.Dispatch.cs:1225]: naming - is this really meant to be named `toDiscordObject`

File: DSharpPlus/DiscordIntents.cs [hunk 2]
```diff
@@ -83,7 +83,7 @@ public enum DiscordIntents
         /// Whether to include guild ban events.
         /// <para>These include <see cref="DiscordClient.GuildBanAdded"/>, and <see cref="DiscordClient.GuildBanRemoved"/>.</para>
         /// </summary>
-        GuildBans = 1 << 2,
+        GuildModeration = 1 << 2,
```
Comment (akiraveliara) [inline DSharpPlus/DiscordIntents.cs:86]: why? and why in this PR?

File: DSharpPlus/Entities/AuditLogParser.cs [hunk 3]
```diff
@@ -0,0 +1,1651 @@
...
+            guild._scheduledEvents[discordEvent.Id] = discordEvent;
+        }
+
+        events = guild._scheduledEvents.ToDictionary(x => x.Key, y => y.Value);
+
+        Dictionary<ulong, DiscordThreadChannel> threads = new();
+        if (uniqueThreads.Any())
+        {
+            threads = uniqueThreads.ToDictionary(xa => xa.Id, xa => xa);
+        }
```
Comment (akiraveliara) [inline DSharpPlus/Entities/AuditLogParser.cs:58]: do we cache threads anywhere? if so, worth updating

File: DSharpPlus/Entities/AuditLogParser.cs [hunk 4]
```diff
@@ -0,0 +1,1651 @@
...
+            case AuditLogActionType.AutoModerationRuleUpdate:
+            case AuditLogActionType.AutoModerationRuleDelete:
+                entry = ParseAutoModerationRuleUpdateEntry(guild, auditLogAction);
+                break;
+
+            default:
+                guild.Discord.Logger.LogWarning(LoggerEvents.AuditLog,
+                    "Unknown audit log action type: {0} - this should be reported to library developers",
+                    (int)auditLogAction.ActionType);
+                break;
```
Comment (akiraveliara) [inline DSharpPlus/Entities/AuditLogParser.cs:417]: this shouldn't be a warning by default - or at least be toggleable, similar to the unknown events warning

File: DSharpPlus/Entities/DiscordAuditLogObjects.cs [hunk 5]
```diff
@@ -23,902 +23,1051 @@
...
+    /// Gets the category under which the action falls.
+    /// </summary>
+    public AuditLogActionCategory ActionCategory { get; internal set; }
+}
+
+/// <summary>
+/// Represents a description of how a property changed.
+/// </summary>
+/// <typeparam name="T">Type of the changed property.</typeparam>
+public struct PropertyChange<T>
```
Comment (akiraveliara) [inline DSharpPlus/Entities/DiscordAuditLogObjects.cs:60]: `readonly record struct`?

File: DSharpPlus/Entities/DiscordAuditLogObjects.cs [hunk 6]
```diff
@@ -23,902 +23,1051 @@
 
 using System;
 using System.Collections.Generic;
+using DSharpPlus.Enums;
 
-namespace DSharpPlus.Entities
+namespace DSharpPlus.Entities;
```
Comment (akiraveliara) [inline DSharpPlus/Entities/DiscordAuditLogObjects.cs:28]: as a principal note, i really dislike how this file contains so many different types, but it's not the end of the world

in an ideal world we'd just sourcegen all this

**Review by akiraveliara — COMMENTED** [review]
left a handful of comments, overall it looks good though

**Review by InFTord — COMMENTED** [review]
more documentation required, but other than that LGTM

File: DSharpPlus/DiscordIntents.cs (hunk 2 above)
Comment (Plerx2493) [reply DSharpPlus/DiscordIntents.cs:86]: It was renamed by Discord but idk why i put it in this PR

File: DSharpPlus/Entities/AuditLogParser.cs (hunk 4 above)
Comment (Plerx2493) [reply DSharpPlus/Entities/AuditLogParser.cs:417]: I think i will make it toggleable

File: DSharpPlus/Entities/DiscordAuditLogObjects.cs (hunk 6 above)
Comment (Plerx2493) [reply DSharpPlus/Entities/DiscordAuditLogObjects.cs:28]: I could move them to own files if its prefered

File: DSharpPlus/Entities/AuditLogParser.cs (hunk 3 above)
Comment (Plerx2493) [reply DSharpPlus/Entities/AuditLogParser.cs]: is updated

**Comment by Plerx2493** [comment]
> may close #1569

I haven't had a chance to check it yet, but I can look at it later.

**Comment by InFTord** [comment]
yes, xmldocs

File: DSharpPlus/Entities/AuditLogParser.cs (hunk 4 above)
Comment (akiraveliara) [reply DSharpPlus/Entities/AuditLogParser.cs:442]: probably a good idea, especially considering we know there's undocumented types

File: DSharpPlus/Entities/DiscordAuditLogObjects.cs (hunk 6 above)
Comment (akiraveliara) [reply DSharpPlus/Entities/DiscordAuditLogObjects.cs:28]: would be, but it's not the end of the world if we keep it here

**Review by akiraveliara — APPROVED** [review]
lgtm, will wait for preliminary test results before merging - if you want to split the types off into their own files, feel free

**Comment by Plerx2493** [comment]
> lgtm, will wait for preliminary test results before merging - if you want to split the types off into their own files, feel free

currently working on it

**Comment by Plerx2493** [comment]
Should be ready now

**Review by InFTord — APPROVED** [review]
LGTM

**Comment by Plerx2493** [comment]
Missing change keys:

DiscordAuditLogAutoModerationRuleEntry:

- [ ] $add_keyword_filter
- [ ] $remove_keyword_filter

DiscordAuditLogIntegrationEntry:

- [ ] type
- [ ] name

DiscordAuditLogThreadEventEntry:

- [ ] flags

**Comment by akiraveliara** [comment]
once the change keys are added we can merge this as far as i'm concerned

**Pull request merged**